Add statx

hasufell opened the pull request:
This is ready for review @hs-viktor @Bodigrim

hasufell commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +718,277 @@ getFdPathVar (Fd fd) v =
+getExtendedFileStatus_ fdMay str StatxFlags{..} StatxMask{..} = do
+  fp <- mallocForeignPtrBytes (#const sizeof(struct statx))
+  withForeignPtr fp $ \p ->
+      throwErrnoIfMinus1Retry_ "getExtendedFileStatus_" (c_statx c_fd str flags masks p)
```
Is `throwErrnoIfMinus1Retry_` good here?

vdukhovni replied:
Are you asking whether it is *necessary* (whether `statx(2)` is interruptible) or whether it is the right thing to do (`statx(2)` is interruptible , but we shouldn't retry)?

The manpage for `statx(2)` does not mention `EINTR`. Plausibly it is not in fact interruptible (caught signals don't interrupt the call).

hasufell commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +718,287 @@ getFdPathVar (Fd fd) v =
+testFlag ex flag =
+  let attributes = fileAttributesX ex
+      attributes_mask = fileAttributesMaskX ex
+  in (attributes .&. attributes_mask .&. flag) == flag
```
is my bit code correct?

vdukhovni reviewed (changes requested):
Preliminary feedback. Let's make *extensive* use of pattern synonyms to avoid constantly changing the API as code point values come and go.

Tying user-visible API datatypes to a point-in-time set of known values is a strong anti-pattern (no pun intended).

Bodigrim commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +718,287 @@ getFdPathVar (Fd fd) v =
+    sec  <- (#peek struct statx, stx_btime.tv_sec) statx_ptr :: IO EpochTime
+    nsec <- (#peek struct statx, stx_btime.tv_nsec) statx_ptr :: IO (#type int)
+    let frac = toInteger nsec % 10^(9::Int)
+    return $ fromRational $ toRational sec + frac
```
`Rational` is likely to be very slow. I'd suggest something like 
```
{-# LANGUAGE NumDecimals #-}
return $ secondsToNominalDiffTime $ MkFixed $ toInteger sec * 1e12 + toInteger nsec * 1e3
```
Could we also please add a test to check that time stamps are interpreted correctly? It's easy to swap `sec` and `nsec`, or treat them as a large `Int64` vs. two `Int32`.

Bodigrim reviewed (commented):
Thanks!

hasufell commented:
All done: tests added, using pattern synonyms now, added more missing functionality.

hasufell commented on tests/FileExtendedStatus.hs:
```diff
@@ -0,0 +1,159 @@
+  ls  <- getExtendedFileStatus Nothing f SymlinkNoFollow defaultStatxMask
+  fs' <- getFileStatus f
+
+  statusExtendedElementsMinimal fs @?= statusElementsMinimal fs'
```
Here we check the intersection of functionality between `stat` and `statx` is equivalent.

hasufell commented on tests/FileExtendedStatus.hs:
```diff
@@ -0,0 +1,159 @@
+{-# OPTIONS_GHC -fno-warn-missing-signatures #-}
+
+module FileExtendedStatus (main) where
```
Mostly a bold copy-pasta of `FileStatus.hs` with some adjustments, but I'm not really bullish of deduplication. Tests are stupid code.

hasufell commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
+
+
+defaultStatxMask :: StatxMask
+defaultStatxMask = StatxAll
```
I set the default mask to `STATX_ALL` here.

hasufell commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
+    return $ timeHiResToNominalDiffTime sec nsec
+
+timeHiResToNominalDiffTime :: EpochTime -> Int32 -> POSIXTime
+timeHiResToNominalDiffTime (CTime sec) nsec = secondsToNominalDiffTime $ MkFixed $ toInteger sec * 1e12 + toInteger nsec * 1e3
```
This is very juicy.

hasufell commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
+fileDaxX _ = error "fileDaxX not available on this platform"
+#endif
+
+fileLinkCountX (ExtendedFileStatus statx) = unsafePerformIO $ do
+  links <- withForeignPtr statx $ (#peek struct statx, stx_nlink)
+  return $ CNlink (fromIntegral (links :: Word32))
```
This is not an accident. We get borked results if we use `CNlink` directly:

```
$ cabal run unix-tests -- -p '/fileExtendedStatus/'                                                                               44.8s  Sat 12:45
All
  Native
    fileExtendedStatus: FAIL
      tests/FileExtendedStatus.hs:79:
      expected: 2
       but got: 4294967296002
```

Definitions are:

- statx
  - C code: `__u32`
  - Haskell type: `Word32`
- stat
  - C code: `nlink_t`
  - Haskell type: `CNlink` (wrapped `Word64`)

I'm not too sure what's going on here @vdukhovni.

hasufell commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
+-- statx
+--
+
+newtype {-# CTYPE "__u64" #-} CAttributes = CAttributes Int64
```
Is `__u64` here safe to use?

hasufell commented:
OpenBSD servers look busted or something

hasufell commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +753,418 @@ getFdPathVar (Fd fd) v =
+    return $ timeHiResToNominalDiffTime sec nsec
+
+timeHiResToNominalDiffTime :: EpochTime -> Int32 -> POSIXTime
+timeHiResToNominalDiffTime (CTime sec) nsec = realToFrac $ picosecondsToDiffTime $ toInteger sec * 1e12 + toInteger nsec * 1e3
```
`secondsToNominalDiffTime` is not available on older `time` versions, so have to resort to `picosecondsToDiffTime` and `realToFrac` (which has a rewrite rule between DiffTime and NominalDiffTime)

Bodigrim replied:
I think it's time we drop support for GHC 8.2 and 8.4, I have not seen anyone using them for ages.

hasufell replied:
Done

vdukhovni commented on System/Posix/Files/Common.hsc:
```diff
@@ -688,3 +753,443 @@ getFdPathVar (Fd fd) v =
+fileImmutableX ex = testFlag ex (#const STATX_ATTR_IMMUTABLE)
+#else
+{-# WARNING fileImmutableX "fileImmutableX: not available on this platform (CPP guard: @#if STATX_ATTR_IMMUTABLE@)" #-}
+fileImmutableX _ = error "fileImmutableX not available on this platform"
```
Here also, nnd in all similar cases. But the compile-time warnings stay.

vdukhovni reviewed (commented):
Thanks for the `PatternSynonym` rewrite. I hope agree this is the better approach.
Looks good overall, but I haven't meticulously read all the code.
I see some remaining `TODO` comments. Are these for this or future PRs?

vdukhovni commented:
It seems there are some missing imports. The FreeBSD test CI fails with:
```
System/Posix/Files/Common.hsc:1193:43: error:
    Variable not in scope: ioeSetLocation :: t0 -> String -> IOError
     |
1193 | getExtendedFileStatus_ _ _ _ _ = ioError (ioeSetLocation unsupportedOperation "getExtendedFileStatus")
     |                                           ^^^^^^^^^^^^^^

System/Posix/Files/Common.hsc:1193:58: error:
    Variable not in scope: unsupportedOperation
     |
1193 | getExtendedFileStatus_ _ _ _ _ = ioError (ioeSetLocation unsupportedOperation "getExtendedFileStatus")
     |                                                          ^^^^^^^^^^^^^^^^^^^^
```

hasufell commented:
> I see some remaining TODO comments. Are these for this or future PRs?

Yeah, it wasn't immediately clear to me how to do them, so they can be added later imo.

hasufell commented on unix.cabal:
```diff
@@ -69,10 +69,10 @@ library
     build-depends:
-        base        >= 4.10    && < 4.20,
-        bytestring  >= 0.9.2   && < 0.13,
+        base        >= 4.12.0.0  && < 4.20,
```
Here we restrict to GHC >= 8.6.x

hasufell commented on unix.cabal:
```diff
@@ -69,10 +69,10 @@ library
+        bytestring  >= 0.9.2     && < 0.13,
         filepath    >= 1.4.100.0 && < 1.5,
-        time        >= 1.2     && < 1.13
+        time        >= 1.9.1     && < 1.13
```
https://hackage.haskell.org/package/time-1.12.2/docs/Data-Time-Clock.html#v:secondsToNominalDiffTime

Bodigrim reviewed (approved):
Looks good to me, although I'm no expert in the area.

hasufell commented:
* made naming more consistent with `fileStatus` variants
* more CPP guards
* added `deviceIDX`, `specialDeviceIDX` and `mountIDX` ...this should be complete now

vdukhovni reviewed (commented):
Finished read through. Just some cosmetic comments/questions.

The tests are a bit skimpy perhaps. They don't cover much of the fancy `statx` functionality, but admittedly tricky to set up in a portable way (may need root privs or specialised filesystems mounted).  It would be nice to cover some of the more portable features.

That said, basically ready to approve.

hasufell commented:
* added missing flags
* removed unnecessary CPP
* massively improved docs

Wrt tests... I think I'll leave them as-is for now. Maybe a follow-up PR can add more.

vdukhovni reviewed (approved):
LGTM
